src/Microsoft.Health.Fhir.Shared.Api/Controllers/EverythingController.cs

```diff
@@ -0,0 +1,126 @@
+    {
+        private readonly IMediator _mediator;
+        private readonly RequestContextAccessor<IFhirRequestContext> _fhirRequestContextAccessor;
+        private static readonly HashSet<string> _supportedParameters = new()
+        {
+            EverythingOperationParameterNames.Start,
+            EverythingOperationParameterNames.End,
+            KnownQueryParameterNames.Since,
+            KnownQueryParameterNames.Type,
+            KnownQueryParameterNames.ContinuationToken,
+        };
+
+        public EverythingController(
+            IMediator mediator,
+            RequestContextAccessor<IFhirRequestContext> fhirRequestContextAccessor)
+        {
+            EnsureArg.IsNotNull(mediator, nameof(mediator));
+            EnsureArg.IsNotNull(fhirRequestContextAccessor, nameof(fhirRequestContextAccessor));
+
+            _mediator = mediator;
+            _fhirRequestContextAccessor = fhirRequestContextAccessor;
+        }
+
+        /// <summary>
+        /// Returns resources defined in $everything operation
+        /// </summary>
+        /// <param name="idParameter">The resource ID</param>
+        [HttpGet]
+        [Route(KnownRoutes.PatientEverythingById, Name = RouteNames.PatientEverythingById)]
+        [AuditEventType(AuditEventSubType.Everything)]
+        public async Task<IActionResult> PatientEverythingById(string idParameter)
+        {
+            ValidateAndGetParameters(out PartialDateTime start, out PartialDateTime end, out PartialDateTime since, out string type, out string ct);
+
+            EverythingOperationResponse result = await _mediator.Send(new EverythingOperationRequest(ResourceType.Patient.ToString(), idParameter, start, end, since, type, ct), HttpContext.RequestAborted);
+
+            return FhirResult.Create(result.Bundle);
+        }
+
+        private void ValidateAndGetParameters(
+            out PartialDateTime start,
+            out PartialDateTime end,
+            out PartialDateTime since,
+            out string type,
+            out string ct)
+        {
+            IReadOnlyList<Tuple<string, string>> inputParameters = Request.Query
+                .SelectMany(query => query.Value, (query, value) => Tuple.Create(query.Key, value))
+                .ToArray();
+
+            start = ReadPartialDateTimeParameter(inputParameters, EverythingOperationParameterNames.Start);
+            end = ReadPartialDateTimeParameter(inputParameters, EverythingOperationParameterNames.End);
+            since = ReadPartialDateTimeParameter(inputParameters, KnownQueryParameterNames.Since);
+            type = ReadStringParameter(inputParameters, KnownQueryParameterNames.Type);
+            ct = ReadStringParameter(inputParameters, KnownQueryParameterNames.ContinuationToken);
+            ProcessUnsupportedInputParameters(inputParameters);
+        }
+
+        private static string ReadStringParameter(IReadOnlyList<Tuple<string, string>> inputParameters, string parameterName)
```

Comment:
Why do we need these parameter reading methods instead of the FromQuery method used in the old version?

Reply:
Right, the valid ones are read from the FromQuery (as in the old version) in the update.